src/lib.rs

```diff
@@ -95,130 +95,165 @@ fn impl_with_generator<G: BodyGenerator>(ast: &syn::DeriveInput, gen: G) -> quot
+                FieldKind::resolve(&field.ty).move_or_clone_field(&index)
+            });
+            quote! { ( #(#fields),* ) }
         }
     }
 
-    fn visit_enum_data(&self, ident: quote::Tokens, data: &syn::VariantData) -> quote::Tokens {
-        match *data {
-            syn::VariantData::Struct(ref body) => {
-                let idents = body.iter()
-                    .map(|field| field.ident.as_ref().unwrap())
-                    .collect::<Vec<_>>();
-                let cloned = body.iter()
-                    .map(|field| {
-                        let ref ident = field.ident.as_ref().unwrap();
-                        let ident = quote! { #ident };
-                        let code = FieldKind::resolve(field).move_or_clone_field(&ident);
-                        quote! { #ident: #code }
-                    })
-                    .collect::<Vec<_>>();
-                quote! { #ident { #(#idents),* } => #ident { #(#cloned),* } }
-            },
-            syn::VariantData::Tuple(ref body) => {
-                let idents = (0..body.len())
-                    .map(|index| syn::Ident::from(format!("x{}", index)))
-                    .collect::<Vec<_>>();
-                let cloned = idents.iter().zip(body.iter())
-                    .map(|(ident, field)| {
-                        let ident = quote! { #ident };
-                        FieldKind::resolve(field).move_or_clone_field(&ident)
-                    })
-                    .collect::<Vec<_>>();
-                quote! { #ident ( #(#idents),* ) => #ident ( #(#cloned),* ) }
-            },
-            syn::VariantData::Unit => {
-                quote! { #ident => #ident }
-            },
+    fn visit_enum_data(
+        &self,
+        ident: proc_macro2::TokenStream,
+        variant: &syn::Variant,
+    ) -> proc_macro2::TokenStream {
+        if variant.fields.is_empty() {
+            return quote!(#ident => #ident);
+        }
+
+        let fields_are_named = variant.fields.iter().any(|field| field.ident.is_some());
+
+        if fields_are_named {
+            let named_fields = variant
+                .fields
+                .iter()
+                .filter_map(|field| field.ident.as_ref());
+
+            let cloned = variant.fields.iter().map(|field| {
+                let ident = field.ident.as_ref().unwrap();
+                let ident = quote!(#ident);
+                let code = FieldKind::resolve(&field.ty).move_or_clone_field(&ident);
+                quote! { #ident: #code }
+            });
+            quote! { #ident { #(#named_fields),* } => #ident { #(#cloned),* } }
+        } else {
+            let abc = |i| ('a'..='z').nth(i);
+            let unnamed_fields = &variant
+                .fields
+                .iter()
+                .enumerate()
+                .filter(|(_, field)| field.ident.is_none())
+                .map(|(index, _)| format_ident!("{}", abc(index).expect("so many indexes")))
```

Comment:
The previous way for ident generation was `x{index}` ... The a..=z is probably enough for everyone, just wondering if you picked this up from some material as a suggested way to generate identifiers? I wasn't able to find google this up.

Reply:
oh that's a good idea actually, I was a bit in creative mood at that moment, but `x{index}` or possibly `arg_{index}` are actually more reasonable, will do that